drivers/gpio/gpio_sim.c

```diff
@@ -0,0 +1,563 @@
+				interrupts |= (1 << i);
+			}
+			break;
+
+		case GPIO_INT_EDGE_BOTH:
+			if (prev_bit != bit) {
+				interrupts |= (1 << i);
+			}
+			break;
+
+		case GPIO_INT_LEVEL_LOW:
+			if (!bit) {
+				interrupts |= (1 << i);
+			}
+			break;
+
+		case GPIO_INT_LEVEL_HIGH:
+			if (bit) {
+				interrupts |= (1 << i);
+			}
+			break;
+
+		case 0:
+		case GPIO_INT_DISABLE:
+			break;
+
+		default:
+			LOG_DBG("unhandled case %u",
+			  config->flags[i] & interrupt_bitmask);
+			break;
+		}
+	}
+
+	config->interrupts |= interrupts;
+
+	if (interrupts) {
+		gpio_fire_callbacks(&config->callbacks, port, interrupts);
+	}
+}
+
+int gpio_sim_input_set_masked(struct device *port, gpio_port_pins_t mask,
+			      gpio_port_value_t values)
+{
+	int r;
+	struct gpio_sim_data *drv_data =
+		(struct gpio_sim_data *)port->driver_data;
+	struct gpio_sim_config *config =
+		(struct gpio_sim_config *)drv_data->config;
+	gpio_port_pins_t input_mask;
+	gpio_port_pins_t prev_values;
+
+	if (mask == 0) {
+		r = 0;
+		goto out;
+	}
+
+	k_mutex_lock(&config->mu, K_FOREVER);
+
+	if (~config->common.port_pin_mask & mask) {
+		r = EINVAL;
+		goto unlock;
+	}
+
+	if (values & ~mask) {
+		r = EINVAL;
+		goto unlock;
+	}
+
+	input_mask = gpio_sim_input_mask(config);
+
+	if (~input_mask & mask) {
+		r = EINVAL;
+		goto unlock;
+	}
+
+	prev_values = config->vals;
+
+	config->vals &= ~mask;
+	config->vals |= values;
+
+	values = config->vals;
+
+	gpio_sim_pend_interrupt(port, mask, prev_values, values);
+
+	r = 0;
+
+unlock:
+	k_mutex_unlock(&config->mu);
+
+out:
+	return r;
+}
+
+int gpio_sim_output_get_masked(struct device *port, gpio_port_pins_t mask,
+			       gpio_port_value_t *values)
+{
+	int r;
+	struct gpio_sim_data *drv_data =
+		(struct gpio_sim_data *)port->driver_data;
+	struct gpio_sim_config *config =
+		(struct gpio_sim_config *)drv_data->config;
+	gpio_port_pins_t output_mask;
+
+	if (mask == 0) {
+		r = 0;
+		goto out;
+	}
+
+	k_mutex_lock(&config->mu, K_FOREVER);
+
+	if (~config->common.port_pin_mask & mask) {
+		r = EINVAL;
+		goto unlock;
+	}
+
+	output_mask = gpio_sim_output_mask(config);
+
+	if (~output_mask & mask) {
+		r = EINVAL;
+		goto unlock;
+	}
+
+	*values = config->vals & mask;
+
+	r = 0;
+
+unlock:
+	k_mutex_unlock(&config->mu);
+
+out:
+	return r;
+}
+
+int gpio_sim_flags_get(struct device *port, gpio_pin_t pin, uint32_t *flags)
+{
+	struct gpio_sim_data *drv_data =
+		(struct gpio_sim_data *)port->driver_data;
+	struct gpio_sim_config *config =
+		(struct gpio_sim_config *)drv_data->config;
+
+	/* TODO: check arguments */
+
+	*flags = config->flags[pin];
+
+	return 0;
+}
+
+/*
+ * GPIO Driver API
+ */
+
+static int gpio_sim_pin_configure(struct device *port, gpio_pin_t pin,
+				  gpio_flags_t flags)
+{
+	int r;
+	struct gpio_sim_data *drv_data =
+		(struct gpio_sim_data *)port->driver_data;
+	struct gpio_sim_config *config =
+		(struct gpio_sim_config *)drv_data->config;
+
+	if (0 == ((1 << pin) & config->common.port_pin_mask)) {
```

Comment:
0 == ((1 << pin)

This seems to be a Zephyr style. The avoidance of ! is really surprising to me. I would write this as

unsigned int mask = BIT(pin);

if (!(config->common.port_pin_mask & mask))

but that seems to not be done in Zephyr? I do think a mask var would help in this function though

Reply:
There's some push towards using MISRA compliant code in Zephyr, and MISRA has a concept of essential types which disallows implicit conversion between integer and boolean values.  Thus you'll often see `x == 0` or `x != 0` in cases where `x` is sufficient from a C language perspective.

Also while I happen to like Yoda comparisons they're not acceptable to `checkpatch`, so it'd be better to avoid `0 == x`.

The above expression should be:

    if ((config->common.port_pin_mask & BIT(pin)) == 0) {